drivers/rtc/rtc_sam.c

```diff
@@ -0,0 +1,681 @@
+	*mask |= rtc_sam_alarm_mask_from_calalr(calalr);
+
+	rtc_sam_tm_from_timalr_calalr(timeptr, *mask, timalr, calalr);
+	return 0;
+}
+
+static int rtc_sam_alarm_is_pending(const struct device *dev, uint16_t id)
+{
+	struct rtc_sam_data *data = dev->data;
+
+	const struct rtc_sam_config *config = dev->config;
+
+	if (id != 0) {
+		return -EINVAL;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	if ((config->regs->RTC_SR & RTC_SR_ALARM) == 0) {
+		k_spin_unlock(&data->lock, key);
+
+		return 0;
+	}
+
+	config->regs->RTC_SCCR = RTC_SCCR_ALRCLR;
+	k_spin_unlock(&data->lock, key);
+	return 1;
+}
+
+static int rtc_sam_alarm_set_callback(const struct device *dev, uint16_t id,
+				      rtc_alarm_callback callback, void *user_data)
+{
+	struct rtc_sam_data *data = dev->data;
+
+	const struct rtc_sam_config *config = dev->config;
+
+	if (id != 0) {
+		return -EINVAL;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	irq_disable(config->irq_num);
+	data->alarm_callback = callback;
+	data->alarm_user_data = user_data;
+
+	if (data->alarm_callback) {
+		config->regs->RTC_IER = RTC_IER_ALREN;
+	} else {
+		config->regs->RTC_IDR = RTC_IDR_ALRDIS;
+	}
+
+	irq_enable(config->irq_num);
+	k_spin_unlock(&data->lock, key);
+	return 0;
+}
+#endif /* CONFIG_RTC_ALARM */
+
+#ifdef CONFIG_RTC_UPDATE
+static int rtc_sam_update_set_callback(const struct device *dev, rtc_update_callback callback,
+				       void *user_data)
+{
+	struct rtc_sam_data *data = dev->data;
+
+	const struct rtc_sam_config *config = dev->config;
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	irq_disable(config->irq_num);
+
+	data->update_callback = callback;
+	data->update_user_data = user_data;
+
+	if (data->update_callback) {
+		config->regs->RTC_IER = RTC_IER_SECEN;
+	} else {
+		config->regs->RTC_IDR = RTC_IDR_SECDIS;
+	}
+
+	irq_enable(config->irq_num);
+
+	k_spin_unlock(&data->lock, key);
+
+	return 0;
+}
+#endif /* CONFIG_RTC_UPDATE */
+
+#ifdef CONFIG_RTC_CALIBRATION
+static int rtc_sam_set_calibration(const struct device *dev, int32_t calibration)
+{
+	struct rtc_sam_data *data = dev->data;
+	const struct rtc_sam_config *config = dev->config;
+	bool negative_calibration;
+	bool high_calibration;
+	uint32_t slow_clock_calibration;
+	uint32_t mr;
+
+	if ((calibration < RTC_SAM_CALIBRATE_PPB_MIN) ||
+	    (calibration > RTC_SAM_CALIBRATE_PPB_MAX)) {
+		return -EINVAL;
+	}
+
+	/* The value written to the register is absolute */
+	if (calibration < 0) {
+		negative_calibration = true;
+		calibration = -calibration;
+	} else {
+		negative_calibration = false;
+	}
+
+	if (calibration < RTC_SAM_CALIBRATE_PPB_QUANTA) {
+		high_calibration = false;
+		slow_clock_calibration = 0;
+	} else if (calibration < RTC_SAM_CALIBRATE_PPB_LOW_SCALE) {
+		high_calibration = false;
+		slow_clock_calibration = (uint32_t)((3906000 / (20 * calibration)) - 1);
+	} else {
+		high_calibration = true;
+		slow_clock_calibration = (uint32_t)((3906000 / calibration) - 1);
+	}
```

Comment:
I would like a comment to state from where this formula comes from:

/* SAMxx datasheet ver. xyz pg. xpto */

define the 3906000 and lets use a name for it